reana_client/utils.py

```diff
@@ -78,7 +78,7 @@ def yadage_load(workflow_file, toplevel=".", **kwargs):
     }
 
     try:
-        yadageschemas.load(
+        return yadageschemas.load(
```

Comment:
I'm not sure if this can have some repercussions, as this operation is done in the workflow-engine level [here](https://github.com/reanahub/reana-workflow-engine-yadage/blob/0cb641bd8b7e233dae6504ebbf8cf8e7091b7176/reana_workflow_engine_yadage/cli.py#L125-L130).

I think here it isn't returning anything because is run uniquely for validation purposes.

Maybe @tiborsimko @diegodelemos you know better?

Reply:
Checking [this PR](https://github.com/reanahub/reana-workflow-engine-yadage/pull/150/files), I think I understand the reason behind moving the `yadadeschemas.load()` call to the workflow-engine.

Yadage has some operation options called `toplevel` and `initdir` that are useful to point out the path to where to look for the workflow file to resolve it. This can admit for instance GitHub URLs (which is pretty cool!). In the past, these options were hardcoded on the workflow-engine level to point always to the current directory. See previous [`toplevel`](https://github.com/reanahub/reana-workflow-engine-yadage/pull/150/files#diff-0e0b04c271e320cd6f4f031cb2acf70a67f1ad43bd8ca6e4b9a8d247e83c0dd1L71) and [`initdir`](https://github.com/reanahub/reana-workflow-engine-yadage/pull/150/files#diff-0e0b04c271e320cd6f4f031cb2acf70a67f1ad43bd8ca6e4b9a8d247e83c0dd1L104) treatment.

As a consequence of [this issue](https://github.com/reanahub/reana-client/issues/396) we were asked to make these options configurable from `reana-client` side, so one could e.g. pass a GitHub URL as a `toplevel` parameter and get the workflow file remotely before executing the workflow.

But you might be wondering... why can't we just do this on the client-side? That's a very good question that I even asked myself half an hour ago.

The reason is that these options are relative paths which [we need to prepend the workflow workspace to](https://github.com/reanahub/reana-workflow-engine-yadage/pull/150/files#diff-0e0b04c271e320cd6f4f031cb2acf70a67f1ad43bd8ca6e4b9a8d247e83c0dd1R47), and the workflow workspace is only available on the cluster once the workflow has been created. All this has to be done before [calling `yadageschemas.load()`](https://github.com/reanahub/reana-workflow-engine-yadage/pull/150/files#diff-0e0b04c271e320cd6f4f031cb2acf70a67f1ad43bd8ca6e4b9a8d247e83c0dd1R109) to have the correct paths.

Said that, I think that after these changes we never pass the Yadage workflow specification as JSON and we always point to the file, to do all the logic on the engine level. Actually, as you can see [here](https://github.com/reanahub/reana-workflow-controller/blob/d1b14ee1aa33c6c7b376b1ce99e49909d3a4db26/reana_workflow_controller/workflow_run_manager.py#L102),  we removed the `--workflow-json` option altogether, in comparison with the other engines.

So I think this means that it might be safe to resolve the workflow spec on the client-side to perform validations, as we will resolve it again in the engine anyway, building the proper operational options.

Reply:
Another update on this, it seems that runtime parameters are respected with this PR checked out. Plus, I had completely forgotten that we show these separately in the Spec tab.

![image](https://user-images.githubusercontent.com/8863238/109782436-2effd480-7c09-11eb-9605-dd759d621270.png)

_Note the "...gendata.C(**1000**,"/var/reana/..." in the logs._

![image](https://user-images.githubusercontent.com/8863238/109782648-65d5ea80-7c09-11eb-9fc2-d8ebb6c96263.png)

Reply:
The main drawback I see is what @diegodelemos mentioned. If you run a workflow with this PR, on will get the Yadage spec visible in the UI. e.g. one will be able to see the Yadage environment image and tag. But if the user re-uploads the yadage/workflow.yaml spec with some changes, those won't be visible in the Spec tab.

Whereas with the previous approach, the yadage workflow spec wasn't visible, instead its path was shown, which could be accessed via the Workspace tab with the latest changes.

Reply:
Thanks @mvidalgarcia for the detailed explanation. 💯 

I will try to conclude this here and please correct me here if I am deviating somewhere.

- If we `return` the `spec` from `yadage_load`, the validation will work fine without any issues.
- The problem will be that It will not be in sync with `r-ui` and `r-server` while running workflows.
- Instead of paths, we will have the `spec` because of `return` and if user do the changes later, it will not be visible or applied in running workflows.

If I am understanding  it correct so far, then we can make a separate method for returning/loading the schema and use it for validation purposes and inform user about (what/which workflow spec) is being the validated.

What do you think?

Reply:
If we decide that we don't want to expose the yadage workflow spec in the UI, one possible solution would be to still load and return yadage spec from [load_workflow_spec()](https://github.com/reanahub/reana-client/blob/03460b791804e287063a7b4e56cb6bf7c7d78e19/reana_client/utils.py#L94) (same interface for all engines will be kept) and once the spec is inside [load_reana_spec()](https://github.com/reanahub/reana-client/blob/03460b791804e287063a7b4e56cb6bf7c7d78e19/reana_client/utils.py#L94) and all the parameter validation is done it can be set to `None` with a comment and a link to this discussion..
Because we still have to return the spec for additional validation. It is also needed for this [PR](https://github.com/reanahub/reana-client/pull/470/files)

Reply:
Use case described, based on `reana-demo-root6-roofit`:

- Checkout this PR
- Run Yadage workflow: `reana-client run -w roofit-yad -f reana-yadage.yaml`
- Check on the UI the spec tab of the workflow, look for `image` and `imagetag`:
  ```console
    "image": "reanahub/reana-env-root6",
    "imagetag": "6.18.04",
  ```
- Modify `workflow/yadage/workflow.yaml` and change the image to an invalid one, e.g. `reanahub/reana-env-root666`.
- Upload the new Yadage spec: `reana-client upload -w roofit-yad workflow/yadage/workflow.yaml`
- Restart the workflow: `reana-client restart -w roofit-yad`
- Workflow fails due to invalid image but in the Spec the old image remains, leaving us in an inconsistent state (meaning that it's also inconsistent in the DB).

Conclusion: the `return` statement has an impact on the consistency of the Spec stored on DB and displayed on UI.


Possible solution: Would it be possible to **only** return the spec for validation purposes and not always?

Reply:
I think it is possible to pass an optional flag (`validation=True`) to `load_reana_spec()` and return on the basis of that flag.

What do you think @mvidalgarcia @audrium ?

Reply:
`load_reana_spec()` already has a `skip_validation` flag, so it wouldn't be clear its meaning. Maybe `yadage_load` can get an optional `kwarg` called `return` and return on basis of that.

Reply:
The problem is that parameters validation is not optional (I'm referring to [this](https://github.com/reanahub/reana-client/issues/386) issue), so while the flag would kind of work in this issue context, it won't for upcoming issues

Reply:
Even if it will be not optional, having a flag to return or not the spec will be helpful in case of sync with `r-ui` and `r-server`.